### PR TITLE
[InstCombine] Fold ((cst << x) & 1) --> x == 0 when cst is odd

### LINKED ISSUE #73384: `((cst << x) & 1)` --> `x == 0` when cst is odd
Test case:
https://godbolt.org/z/15x6fe7vr
```cpp
int f3(int c)
{
  return ((3 << c) & 1);
}
```

See Also: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96669

### PR DESCRIPTION
**Pull request by elhewaty**
- [InstCombine] Add test coverage ((cst << x) & 1) --> x == 0 when cst is odd (NFC)
- [InstCombine] Fold ((cst << x) & 1) to x == 0 when cst is odd
- Fixes: https://github.com/llvm/llvm-project/issues/73384
- Alive2: https://alive2.llvm.org/ce/z/5RbaK6

### CODE REVIEW

**Comment by elhewaty** [comment]
@dtcxzyw @goldsteinn @arsenm

**Comment by elhewaty** [comment]
@dtcxzyw, Can you please take a look here?

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 1]
```diff
@@ -2220,10 +2220,13 @@ Instruction *InstCombinerImpl::visitAnd(BinaryOperator &I) {
   Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
 
   Value *X, *Y;
-  if (match(Op0, m_OneUse(m_LogicalShift(m_One(), m_Value(X)))) &&
+  const APInt *C;
+  if ((match(Op0, m_OneUse(m_LogicalShift(m_One(), m_Value(X)))) ||
+       (match(Op0, m_OneUse(m_Shl(m_APInt(C), m_Value(X)))) &&
+        C->countr_zero() == 0)) &&
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:2226]:
```suggestion
        (*C)[0])) &&
```

File: llvm/test/Transforms/InstCombine/and.ll [hunk 2]
```diff
@@ -7,6 +7,105 @@ declare void @use32(i32)
...
+; CHECK-NEXT:    [[AND:%.*]] = zext <2 x i1> [[TMP1]] to <2 x i32>
+; CHECK-NEXT:    ret <2 x i32> [[AND]]
+;
+  %shl = shl <2 x i32> <i32 5, i32 5>, %x
+  %and = and <2 x i32> %shl, <i32 1, i32 1>
+  ret <2 x i32> %and
+}
+
+; help wanted: why the optimization does not work on this case
+; or when vector elements are different, even if they are odd numbers?
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/and.ll:75]: `m_APInt` only captures a constant scalar or a constant splat (i.e., all elements are the same).

You can handle this case with `KnownBits`. But I don't think it is worth doing that.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/test/Transforms/InstCombine/and.ll (hunk 2 above)
Comment (elhewaty) [reply llvm/test/Transforms/InstCombine/and.ll:75]: @dtcxzyw, why did the build fail?

Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/and.ll:75]: > @dtcxzyw, why did the build fail?

Don't worry about the build failure on windows. It is not related to your patch.

**Comment by elhewaty** [comment]
@dtcxzyw, @nikic ping

**Pull request merged**